Windows containers support

Elufimov opened the pull request:
- [x] New parameters in agent configuration:
docker.client.version
docker.client.os.arch
docker.server.version
docker.server.os.arch
I saved `docker` parameter for backward compatibility.
- [x] New shells `cmd.exe` and `powershell.exe`
- [x] Detection windows based containers 
- [x] Chown in windows based containers 
- [ ] Testing vagrant for regressions 
- [x] Testing in Windows server 2016 windows containers
- [x] Testing in Windows server 2016 linux containers
~~- [ ] Testing in Windows server hyper-v containers~~ (Outside of this pull request)
- [x] Testing in Windows 10 windows containers
- [x] Testing in Windows 10 linux containers
- [x] Testing in Docker for Mac
- [x] Testing in Ubuntu for regressions

Elufimov committed:
push docker server and client versions to agent parameters

Elufimov committed:
constants for new agent parameters

Elufimov committed:
new shells, and working windows docker

Elufimov commented:
@jonnyzzz do we need chown in windows based container? Right now this step looks like 
<img width="735" alt="aa37fca0b942fa76ea8d4b359bfa4e0a" src="https://cloud.githubusercontent.com/assets/423130/22625692/d25bf2f4-ebb6-11e6-80bb-3257e6d5f832.png">

jonnyzzz commented:
@Elufimov we added chown to fix permissions on Linux containers. The fact is/was that all files that were created by the container were owned by root on the host. So build agent is unable to remove those files and it eventually it consumes whole free space with such files.

On macos there is no such issue.

jonnyzzz commented:
I'd like to suggest to rename `docker.server.*` variables into `docker.host.*` to make them correspond to `DOCKER_HOST` environment variable. Also, it might be helpful to indicate if we have `DOCKER_HOST` variable configured for the agent.

There is also a problem with such parameters, since it is possible to change their value by settings up custom `DOCKER_HOST` environment variable. Overall it mays we shall respect custom `DOCKER_HOST` environment variable in compatibility between runner and agent checks.

jonnyzzz commented:
What it the reason to have `docker.*.os.arch`? How one could use it? 

Is that true that `docker.client.os.arch` === `agent.os.arch`, aka is the same as agent OS and version?

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
 
       @Nullable
-      private String executeCommandWithShell(@NotNull final String name, @NotNull final String... command) {
+      private String[] executeCommandWithShell(@NotNull final String name, @NotNull final String... command) {
```
It's better to return `String` as it used to be and process the output carefully around docker detection

Elufimov replied:
Not relevant after this [commit](https://github.com/jonnyzzz/TeamCity.Virtual/pull/51/commits/7da68b403fe776c8f829420c75889f9981a73f65)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -104,7 +129,7 @@ private String executeCommandWithShell(@NotNull final String name, @NotNull fina
         }
 
-        return result.getStdout();
+        return result.getOutLines();
```
The trick behind `ExecResult#getOutLines` is that it splits the `ExecResult#getStdout` by newlines. This code may yield empty entries on the array or white-space only entries.

Let's return the whole output here (as it was before the change) and process it closer to each of commands. This will allow to avoid any changes around vagrant detection at all

Elufimov replied:
Not relevant after this [commit](https://github.com/jonnyzzz/TeamCity.Virtual/pull/51/commits/7da68b403fe776c8f829420c75889f9981a73f65)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
+        for (int i = 0; i < output.length; i++)
+          cleanedArray[i] = output[i].trim();
+
+        int serverPosition = Arrays.asList(cleanedArray).indexOf("Server:");
```
Can you simply have a regex to search all necessary parameter values?

It looks like `stdout.split("Server:")` and  `stdout.split("Client:")` will be enough to select client and server blocks. And next a regex could be used to parse the rest.

We shall have the implementation as independent from command output as possible. Docker will change, and out code is likely to work without further changes

Elufimov replied:
I just found better (maybe) solution https://docs.docker.com/engine/reference/commandline/version/#/display-docker-version-information
we can get versions without parsing output.

Elufimov replied:
Not relevant after this [commit](https://github.com/jonnyzzz/TeamCity.Virtual/pull/51/commits/7da68b403fe776c8f829420c75889f9981a73f65)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
-        String ver = output.toLowerCase().trim();
-        ver = ver.replaceAll("\\s*docker\\s+version\\s+", "");
-        ver = ver.replaceAll(",?\\s+build\\s+", "-");
+        String[] cleanedArray = new String[output.length];
```
Do you use this array? If yes, please use `List<String>`. This will improve readability of the code

Elufimov replied:
Not relevant after this [commit](https://github.com/jonnyzzz/TeamCity.Virtual/pull/51/commits/7da68b403fe776c8f829420c75889f9981a73f65)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -46,44 +47,68 @@ public void beforeAgentConfigurationLoaded(@NotNull BuildAgent agent) {
-        if (StringUtil.isEmptyOrSpaces(ver)) {
-          LOG.warn("Failed to parse docker version: " + output);
+        for (int i = 0; i < output.length; i++)
+          cleanedArray[i] = output[i].trim();
```
Line can be empty, shall we drop it?

Elufimov replied:
Not relevant after this [commit](https://github.com/jonnyzzz/TeamCity.Virtual/pull/51/commits/7da68b403fe776c8f829420c75889f9981a73f65)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/VMRunnerContext.java:
```diff
@@ -68,14 +69,20 @@ public File getAgentTempDirectory() {
   @NotNull
   public String getCheckoutMountPoint() {
+    String defaultMountPoint;
+    if(SystemInfo.isUnix) {
```
Let's check for `SystemInfo.isWindows` instead. In that case it's clear a path should be `c:\...`

Elufimov replied:
fixed

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/VMRunnerContext.java:
```diff
@@ -68,14 +69,20 @@ public File getAgentTempDirectory() {
 
   @NotNull
   public String getCheckoutMountPoint() {
+    String defaultMountPoint;
```
a `final` declaration is nicer here (can be also combined with ternary operator)

Elufimov replied:
fixed

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -16,13 +16,18 @@
+import com.intellij.openapi.util.SystemInfo;
 import com.jonnyzzz.teamcity.virtual.VMConstants;
 import com.jonnyzzz.teamcity.virtual.run.VMRunnerContext;
+import com.oracle.tools.packager.Log;
```
looks like a wrong import

Elufimov replied:
fixed

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -37,4 +42,32 @@ public String getImageName() throws RunBuildException {
+
+  @NotNull
+  public String getShellLocationInsideContainer() throws RunBuildException {
+    String defaultShellLocation;
```
`final` here too would look nicer

Elufimov replied:
fixed

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -16,13 +16,18 @@
 
 package com.jonnyzzz.teamcity.virtual.run.docker;
 
+import com.intellij.openapi.util.SystemInfo;
```
It is more reliable to use `jetbrains.buildServer.agent.BuildAgentConfiguration#getSystemInfo` to get agent OS instead. Currently you are using semi-deprecated API

Elufimov replied:
fixed

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerVM.java:
```diff
@@ -147,7 +163,7 @@ public BuildProcess startImpl() throws RunBuildException {
                 "-v",
                 checkoutDir.getPath() + ":" + mountPoint + ":" + ctx.getDockerMountMode(),
-                "--workdir=" + mountPoint + "/" + RelativePaths.resolveRelativePath(checkoutDir, workDir),
+                "--workdir=" + mountPoint + ctx.getPathSeparatorInsideContainer() + RelativePaths.resolveRelativePath(checkoutDir, workDir),
```
`RelativePaths#resolveRelativePath` uses `/` as path separators. Let's replace path separators of necessary

same for `mountPoint`

Elufimov replied:
I did't test `RelativePaths#resolveRelativePath` on windows. I will look inside.

jonnyzzz replied:
It may easily turn out we run on Linux to start remote container on Windows. We cannot relay on slashes returned from the command

Elufimov replied:
How we can to use a `volume` in case of remote docker demon?

jonnyzzz replied:
Let's check how it would work with Docker. If not we should warn we cannot use it in the build and and/or in the runner settings UI

Elufimov replied:
It works. I added windows specifics.

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerVM.java:
```diff
@@ -125,12 +139,14 @@ public BuildProcess startImpl() throws RunBuildException {
+        if(ctx.isDockerServerWindowsBased() && ctx.getShellLocationInsideContainer().equals("cmd.exe")) {
+          return  Arrays.asList(ctx.getShellLocationInsideContainer(), "/C", "call " + script.getName());
+        } else if(ctx.isDockerServerWindowsBased() && ctx.getShellLocationInsideContainer().equals("powershell.exe")) {
+          return  Arrays.asList(ctx.getShellLocationInsideContainer(), "-Command", "get-content " + script.getName() + " | Invoke-Expression");
```
As far as I remember, it's more stable to call script as a file instead of doing `Invoke-Expression`. The problem was with missing exit code as in script mode it was not failing asap. Could you please check issues for powershell runner in TeamCity tracker for details. This should be an opensource plugin, btw.

Elufimov replied:
My first intention was to use `powershell.exe -File` but all files are creating with .bat extension. There is a TODO about it https://github.com/jonnyzzz/TeamCity.Virtual/blob/master/agent/src/com/jonnyzzz/teamcity/virtual/run/ScriptFile.java#L81
and I thought this workaround will fit. I will look to plugin to find out how to handle it properly.

jonnyzzz replied:
Maybe it's better to allow plugin users to call powershell instead?

Elufimov replied:
Yes, I agree. User can call powershell.exe from cmd like he can call powershell from bash. Powershell is cross platform for a while already.

jonnyzzz reviewed (changes requested):
Thank you so much for the pull request. Looking forward to merging it

Elufimov commented:
I am going to update this pull request during the next week. Thx for comment, I will try to resolve all of it.

Elufimov committed:
Rework docker detection.
1) Switching to native ability of docker cli to formating output and selecting part of output output
2) Reducing number of published parametrs
3) Small refactioring in DockerContext file

Elufimov committed:
sone refactoring in VMRunnerContext file

Elufimov commented:
About permission restoring. It also don't work on mac os. I did an simple experiment.
1) Start container
win `docker run -v c:\:c:\temp -it microsoft/nanoserver:10.0.14393.693 cmd.exe`
macos `docker run -it -v /Users/elufimov/Documents/Projects/Temp/:/temp ubuntu /bin/bash`
2) Create file in volume 
win `echo 1 > C:\ temp\1.txt`
macos `echo 1 > /temp/1.txt`
3) Check permission
windows
```
$ icacls c:\1.txt
c:\1.txt NT AUTHORITY\SYSTEM:(I)(F)
         BUILTIN\Administrators:(I)(F)
         BUILTIN\Users:(I)(RX)
         S-1-5-93-2-1:(I)(F)

Successfully processed 1 files; Failed processing 0 files
```
macos 
```
 elufimov@MacBook-Pro-Michael  ~  ls -l /Users/elufimov/Documents/Projects/Temp/1.txt
-rw-r--r--  1 elufimov  staff  2 Feb 19 13:57 /Users/elufimov/Documents/Projects/Temp/1.txt
```
I can edit and delete those files from hosts systems.

Elufimov committed:
make fixing chown linux specific and remove powershell shell

Elufimov committed:
add windows specifics to RelativePaths with backward compatibility for vagrant runner

Elufimov commented:
Another experiment.
Windows 10 pro and Docker for Windows 1.3.1
1) Open two powershell terminals
2) Run `docker -H npipe:////./pipe/docker_engine_windows run -it microsoft/nanoserver:10.0.14393.693 cmd.exe` in one of them
3) Run `docker -H npipe:////./pipe/docker_engine run -it ubuntu /bin/bash` in another one in the same time
4) They both start correctly

I did not find any official documentation on this feature. This mean that it can changes in any moment. But it opens great for people who use windows 10/windows server 2016 as build agents. What do you think about that?

jonnyzzz commented:
A great thing. I would concentrate on most trivial solution first to get
more feedback from plugin users. We may have this as optional feature for
sure. Still having in mind this is not documented behavior we have risks
the plugin is broken suddenly

On Feb 23, 2017 10:01, "Michael Elufimov" <notifications@github.com> wrote:

> Another experiment.
> Windows 10 pro and Docker for Windows 1.3.1
>
>    1. Open two powershell terminals
>    2. Run docker -H npipe:////./pipe/docker_engine_windows run -it
>    microsoft/nanoserver:10.0.14393.693 cmd.exe in one of them
>    3. Run docker -H npipe:////./pipe/docker_engine run -it ubuntu
>    /bin/bash in another one in the same time
>    4. They both start correctly
>
> I did not find any official documentation on this feature. This mean that
> it can changes in any moment. But it opens great for people who use windows
> 10/windows server 2016 as build agents. What do you think about that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jonnyzzz/TeamCity.Virtual/pull/51#issuecomment-281934175>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPpryyG2FMYjFoCNJ5c2xTS48ZnJu1Bks5rfUrPgaJpZM4L3d9w>
> .
>

Elufimov commented:
Yes, of course. If you do not have new comments about code itself I will test rest of the cases.

jonnyzzz commented:
Please do. I'll review new changes too :)

On Feb 23, 2017 10:12, "Michael Elufimov" <notifications@github.com> wrote:

> Yes, of course. If you do not have new comments about code itself I will
> test rest of the cases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jonnyzzz/TeamCity.Virtual/pull/51#issuecomment-281936674>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPpr91xQvi2meJmig8BF_HKx2HJw343ks5rfU1sgaJpZM4L3d9w>
> .
>

Elufimov committed:
fix case Windows based agent and linux based docker

Elufimov commented:
I think I have tested all docker cases. Can you help me with testing of vagrant regression? I do not use it in production actually.

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -61,26 +61,30 @@ private void detectVagrant(@NotNull final BuildAgentConfiguration config) {
-        String ver = output.toLowerCase().trim();
-        ver = ver.replaceAll("\\s*docker\\s+version\\s+", "");
-        ver = ver.replaceAll(",?\\s+build\\s+", "-");
+        final String dockerVersion = executeCommandWithShell("docker", "docker version --format {{.Server.Version}}");
```
`docker version --format`` if relatively new key. I did experiment and it showed Docker 1.7.1 does not support this option. Currently, it would mean we no longer support older docker. 
Do we really need this limitation?

Elufimov replied:
I will rewrite this place for better compatibility.

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/VMDetector.java:
```diff
@@ -61,26 +61,30 @@ private void detectVagrant(@NotNull final BuildAgentConfiguration config) {
+      private String cleanOutput(String rawOutput) {
+        return rawOutput
+                .trim()
+                .replace(System.getProperty("line.separator"), "");
```
I would recommend first replace regex  `[\s\r\n]+` with single a whitespace, next to trim. Currently, for multiline output, it's possible to contact lines without spaces at all.

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/RelativePaths.java:
```diff
@@ -26,13 +27,23 @@
-    result = result.replace('\\', '/');
-
-    while (result.endsWith(".") || result.endsWith("/")) result = result.substring(0, result.length() - 1);
+    if(convertToLinux)
```
It looks like we do need similar conversion for `!linux` case as well (e.g. if the code is running on Windows targeting Linux containers)

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/RelativePaths.java:
```diff
@@ -26,13 +27,23 @@
-    while (result.endsWith(".") || result.endsWith("/")) result = result.substring(0, result.length() - 1);
+    if(convertToLinux)
+      result = result.replace('\\', '/');
+    if(isWindows) {
```
Shall we avoid the if and simply trim all possible chars: `/`,  `\`, `.`

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/RelativePaths.java:
```diff
@@ -26,13 +27,23 @@
+
+  @NotNull
+  public static String resolveRelativePath(@NotNull final File baseDir, @NotNull final File path) {
+    return resolveRelativePath(baseDir, path, SystemInfo.isWindows, true);
```
A bit correct is to use AgentConfiguration to get information on whether it's Linux or Windows. SystemInfo is meanwhile in the API, but not the best way to use

Elufimov replied:
I saw using `SystemInfo` in plugin code and just did the same. I will look to `AgentConfiguration`

jonnyzzz replied:
I mean this one `jetbrains.buildServer.agent.BuildAgentConfiguration#getSystemInfo`. An instance of `BuildAgentConfiguration` is available via Spring autowiring, so you may have an instance from `DockerVM` contructor

Maybe it's too complicated, but I see it a nice idea to check the OS once somewhere on the `DockerContext` instance creation

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/VMRunnerContext.java:
```diff
@@ -16,6 +16,7 @@
 
 package com.jonnyzzz.teamcity.virtual.run;
 
+import com.intellij.openapi.util.SystemInfo;
```
Same here. A bit correct is to use AgentConfiguration to get information on whether it's Linux or Windows. SystemInfo is meanwhile in the API, but not the best way to use

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/VMRunnerContext.java:
```diff
@@ -68,22 +69,23 @@ public File getAgentTempDirectory() {
   public String getShellLocation() {
-    String loc = myContext.getRunnerParameters().get(VMConstants.PARAMETER_SHELL);
-    if (loc == null) {
+    final String loc = myContext.getRunnerParameters().get(VMConstants.PARAMETER_SHELL);
```
this variable can be null. The check below is dangerous.

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/VMRunnerContext.java:
```diff
@@ -68,22 +69,23 @@ public File getAgentTempDirectory() {
-    if (loc == null) {
+    final String loc = myContext.getRunnerParameters().get(VMConstants.PARAMETER_SHELL);
+    if (loc.equals("default")) {
       return "/bin/bash";
```
Shall we have an alternative default value for windows container?

Elufimov replied:
I left here vagrant implementation. Docker's one lie in method `isDockerServerWindowsBased()`. Shall I merge them?

jonnyzzz replied:
In the code we have `VMRunnerContext` and tho child classes `DockerContext` and `VagrantContext`. Occasionally, we have a change e0bcf7ff220d1bd61a74e1fb86a7224dd04f6cc2 that added `getDockerMountMode` into the base class, and not into the `DockerContext`. 

I guess the best way to solve it is to move older methods to the `VagrantContext` and add new methods to the `DockerContext`.

It looks nice if we have a clear separation between Host and Container OS preferences (e.g. newlines, shell, paths). Currently we are lacking that, as you see

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -37,4 +37,30 @@ public String getImageName() throws RunBuildException {
+
+  public boolean isDockerServerWindowsBased() throws RunBuildException {
+    final String serverArch = myContext.getConfigParameters().get(VMConstants.DOCKER_HOST_OS_PROPERTY);
+    return serverArch.equals("windows");
```
A bit safe is `"windows".equalsIgnoreCase(serverArch)`. 
Does it make sense to look for substring containing windows here?

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -37,4 +37,30 @@ public String getImageName() throws RunBuildException {
+  }
+
+  @NotNull
+  public String getShellLocationInsideContainer() throws RunBuildException {
```
How does `getShellLocation` and the method related to each other? Looks like a code duplicate

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerContext.java:
```diff
@@ -37,4 +37,30 @@ public String getImageName() throws RunBuildException {
+  }
+
+  @NotNull
+  public String getCheckoutMountPointInsideContainer() throws RunBuildException {
```
There is similar looking method `getCheckoutMountPoint`. Shall we merge them?

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerVM.java:
```diff
@@ -88,49 +89,62 @@ public void buildWithScriptFile(@NotNull final File script) throws RunBuildExcep
+        if (SystemInfo.isUnix) {
+          terminatingCommand = Arrays.asList("docker", "kill", name, "2>&1", "||", "true");
+        } else {
+          terminatingCommand = Arrays.asList("docker", "kill", name, ">nul", "2>nul", "&", "exit", "0");
```
This approach swallows possible error output from stop command. It may turn out to be tricky to investigate build issues. Let's redirect STDERR to STDOUT instead.
Is it "&" and not "||"?

jonnyzzz commented on agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerVM.java:
```diff
@@ -88,49 +89,62 @@ public void buildWithScriptFile(@NotNull final File script) throws RunBuildExcep
-                "\"source " + script.getName() + "\""
-        );
+      private List<String> scriptRun(@NotNull final File script) throws RunBuildException {
+        if(ctx.isDockerServerWindowsBased() && ctx.getShellLocationInsideContainer().equals("cmd.exe")) {
```
Would it work for Windows & Powershell?

Elufimov replied:
I tested calling `powershell` from `cmd`. It works.

jonnyzzz commented on server/resources/vm-edit.jsp:
```diff
@@ -46,9 +46,14 @@
+        <props:option value="default">default</props:option>
+        <props:option value="cmd.exe">cmd.exe</props:option>
         <props:option value="/bin/bash">/bin/bash</props:option>
         <props:option value="/bin/sh">/bin/sh</props:option>
```
Not sure if we support /bin/sh

Elufimov replied:
It works.

jonnyzzz reviewed (changes requested):
Thank you so much for improvements you've done to support Windows containers. It looks great!
Nice, you tested it too. 
Changes look entirely local, and I do not expect Vagrant to be broken.

In the code, I found and commented several potential problems. The first one is you use SystemInfo.isWindows and similar methods in the number of places. To me, it'd be better to explicitly define something `getHostConfig()`, `getContainerConfig()`methods near context and use returned values for all decisions and path separators. 

The second thing that there are similarly looking methods in the implementation. For the future bug fixing and code stability, it's better to avoid duplication (if possible)

jonnyzzz commented:
There is one more thing to check: 
Suppose one has an installation of TeamCity with Windows/Linux agents with Docker. Overall 3 possible combinations: Host={Windows} x Container={Windows;Linux} + Host={Linux} x Container={Linux}.

Next, suppose the older version (Linux-Linux only) of the plugin was installed. Say one upgrades the plugin to one supporting Windows too. We must make sure their builds will still be running only on Linux agents, and their agent requirements are not broken too.

The is backward compatibility requirement. There are huge TeamCity setups with hundreds of build configurations and it's impossible to fix side-effects of a plugin update.

VolleMelk commented:
He guys, how are things going? Looks like you are almost there! Keep up the great work!

If you'd like some testing drop, feel free to release an alpha or beta release

gitfool commented:
I'm also interested to take this for a spin using Docker for Windows.

vssubs commented:
Is this dead? /complete ?

jonnyzzz commented:
Many similar features of this plugin are now available out of the box in
TeamCity. This plugin is not actively developed or supported right now. You
may do a pull request if you like or fork it entirely.
